### DISCUSSION TITLE
Question: Is there built in ToUpper() functionality for System.Guid and other value types ?

### DISCUSSION THREAD

**Discussion by TrayanZapryanov** [dotnet/runtime#97795] [Q&A]
Today I was a bit disappointed that there is no built in way to convert guid in upper case. Or maybe I am wrong and it exists...
And wondering if this can be done in a less allocative way - here it is what I produced as extension. 

```
public static class ToUpperExtensions
{
	public readonly struct UpperGuid(Guid data) : ISpanFormattable
	{
		public string ToString(string? format, IFormatProvider? formatProvider)
		{
			Span<char> destination = stackalloc char[68];
			if (!data.TryFormat(destination, out var charsWritten, format))
				return data.ToString(format, formatProvider).ToUpperInvariant();

			destination = destination[..charsWritten];
			if (Ascii.ToUpperInPlace(destination, out _) != OperationStatus.Done)
				return data.ToString(format, formatProvider).ToUpperInvariant();

			return destination.ToString();
		}

		public bool TryFormat(Span<char> destination, out int charsWritten, ReadOnlySpan<char> format, IFormatProvider? provider)
		{
			if (!data.TryFormat(destination, out charsWritten, format))
				return false;

			return Ascii.ToUpperInPlace(destination[..charsWritten], out _) == OperationStatus.Done;
		}
	}

	public static UpperGuid ToUpper(this Guid g) => new(g);
}
``` 

and usage :
```
Console.WriteLine($"Hello, World: {Guid.NewGuid().ToUpper():x}");
Console.WriteLine($"Hello, World: {Guid.NewGuid().ToUpper():n}");
Console.WriteLine($"Hello, World: {Guid.NewGuid().ToUpper():b}");
Console.WriteLine($"Hello, World: {Guid.NewGuid().ToUpper():p}");
Console.WriteLine($"Hello, World: {Guid.NewGuid().ToUpper():d}");
```
Produces:
Hello, World: {0XCAC5A108,0X1B0F,0X49EC,{0XA5,0XCB,0X16,0X90,0X33,0X32,0XF5,0X4B}}
Hello, World: 1EA856EC4737462890D85AA9398FC7E1
Hello, World: {AC5A2F0B-8C7A-4135-8174-86F371E46D56}
Hello, World: (71309472-FA76-47F2-A22F-B1F8C7503A9A)
Hello, World: 3E41F3BA-D001-4F11-AEA2-9A1F3D56674D

Do you think if there are more effective way of doing this?
Also is this a valuable pattern to be used for other types like Date/DateTime..... ?

**Comment by EgorBo** [comment]
It would be nice for format specifiers to work like this (and I think for some types they do work like that):
```
guid.ToString("N") - upper case
guid.ToString("n") - lower case
```

but it'd be a breaking change to introduce this now in `Guid`. If you need ToUpperString, you can do:
```cs
void ToUpper(Guid guid, Span<char> dest, ReadOnlySpan<char> format)
{
    guid.TryFormat(dest, out var charsWritten, format);
    Ascii.ToUpperInPlace(dest.Slice(0, charsWritten), out _);
}
```

**Comment by TrayanZapryanov** [comment]
@EgorBo Thank you for the proposal. I've updated the extension method to reflect it.
Looks like this extension can give some nice allocation benefits in interpolated strings.
Here it is a small benchmark:

```
[MemoryDiagnoser]
public class ToUpperBenchmarks
{

	[Benchmark(Baseline = true)]
	public string ToString_Default_Interpolated()
	{
		return $"Hello, World: {Guid.Empty.ToString().ToUpperInvariant()}";
	}

	[Benchmark]
	public string ToString_ToUpper_Interpolated()
	{
		return $"Hello, World: {Guid.Empty.ToUpper()}";
	}
}
```

BenchmarkDotNet v0.13.10, Windows 11 (10.0.22631.3085/23H2/2023Update/SunValley3)
11th Gen Intel Core i9-11900K 3.50GHz, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.101
  [Host]     : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2


| Method                        | Mean     | Error    | StdDev   | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------------------ |---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| ToString_Default_Interpolated | 35.46 ns | 0.745 ns | 1.784 ns |  1.00 |    0.00 | 0.0268 |     224 B |        1.00 |
| ToString_ToUpper_Interpolated | 52.87 ns | 1.092 ns | 3.079 ns |  1.47 |    0.13 | 0.0153 |     128 B |        0.57 |
